File: 1019.链表中的下一个更大节点.java
/*
 * @lc app=leetcode.cn id=1019 lang=java
 *
 * [1019] 链表中的下一个更大节点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public int[] nextLargerNodes(ListNode head) {
        Stack stack = new Stack();
        ArrayList<Integer> data = new ArrayList<>();
        ArrayList<Integer> res = new ArrayList<>();

        int index = 0;
        while(head != null) {
            res.add(0);
            data.add(head.val);

            while (!stack.empty() && head.val > data.get((Integer)stack.peek())) {
                res.set((Integer)stack.pop(), head.val);
            }

            stack.push(index);
            index++;
            head = head.next;
        }

        int[] ans = new int[res.size()];

        for (int i = 0; i < res.size(); i++) {
            ans[i] = res.get(i);
        }

        return ans;
    }
}

// @lc code=end

File: 1019.链表中的下一个更大节点.cs
/*
 * @lc app=leetcode.cn id=1019 lang=csharp
 *
 * [1019] 链表中的下一个更大节点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int[] NextLargerNodes(ListNode head) {
        var lst = new List<int>();
        var res = new List<int>();
        var st = new Stack<int>();
        ListNode p = head;
        int index = 0;
        while(p!=null)
        {
            res.Add(0);
            lst.Add(p.val);
            while(st.Count!=0 && lst[st.Peek()] < p.val)
            {
                res[st.Peek()] = p.val;
                st.Pop(); 
            }
            st.Push(index++);
        }
        var ans = new int[index];
        for(int i=0;i<index;i++) ans[i] = res[i];
        return ans;
    }
}
// @lc code=end

